lib/ansible/utils/role_finder.py

```diff
@@ -0,0 +1,181 @@
+# (c) 2021 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+
+from ansible.constants import config
+from ansible.parsing.dataloader import path_exists
+from ansible.template import Templar
+from ansible.utils.collection_loader import AnsibleCollectionRef
+from ansible.utils.collection_loader._collection_finder import _get_collection_role_path
+from ansible.utils.path import unfrackpath
+
+
+class AnsibleRoleFinder(object):
+    """
+    Class used to locate the path to any available roles.
+
+    Example usage:
+
+        finder = AnsibleRoleFinder(pb_basedir)
+
+        # Find 'myrole' standard role
+        result = finder.find_first('myrole')
+
+        # Find 'myrole' in ansible.example collection
+        result = finder.find_first('ansible.example.myrole')
+
+        # Find 'myrole' in any of the listed collections or as standard role.
+        result = finder.find_first('myrole', ['ansible.example', 'community.general'])
+
+    TODO:
+      - This would replace code in `ansible-doc` and `ansible-galaxy`. Maybe plugins?
+      - API to filter based on file existence (e.g. meta/main.yml for a-g).
+      - API to show masked roles
+      - API to show only first match
+    """
+
+    def __init__(self, playbook_basedir, role_basedir=None, templar=None):
+        """
+        Initialize a new RoleFinder object.
+
+        :param str playbook_basedir: Path to the playbook base directory.
+        :param str role_basedir: Path to a relative role base directory.
```

Comment:
i don't think we want roles in roles

Reply:
Carry over from `RoleDefinition` code. I'm not entirely sure what this represents there yet, so just carrying the value over until I can explore it some more.